docs/design/expansion.md

```diff
@@ -0,0 +1,133 @@
+
+### Use Case: Composition of environment variables
+
+Currently, containers are inject with docker-style environment variables for the services in their
+pod's namespace.  There are several variables for each service, but users routinely need to compose
+URLs based on these variables because there is not a variable for the exact format they need. 
+Users should be able to build new environment variables with the exact format they need.
+Eventually, it should also be possible to turn off the automatic injection of the docker-style
+variables into pods and let the users consume the exact information they need via the downward API
+and composition.
+
+### Use Case: Variable expansion in command
+
+Users frequently need to pass the values of environment variables to a container's command.  
+Currently, Kubernetes does not perform any expansion of varibles.  The workaround is to invoke a
+shell in the container's command nd have the shell perform the substitution, or to write a wrapper
+script that sets up the environment and runs the command.  This has a number of drawbacks:
+
+1.  Solutions that require a shell are unfriendly to images that do not contain a shell
+2.  Wrapper scripts make it harder to use images as base images
+3.  Wrapper scripts increase coupling to kubernetes
+
+Users should be able to do the 80% case of variable expansion in command without writing a wrapper
+script or adding a shell invocation to their containers' commands.
+
+### Use Case: Images without shells
+
+The current workaround for variable expansion in a container's command requires the container's
+image to have a shell.  This is unfriendly to images that do not contain a shell (`scratch` images,
+for example).  Users should be able to perform the other use-cases in this design without regard to
+the content of their images.
+
+## Design Considerations
+
+### What features should be supported?
+
+In order to limit complexity, we want to provide the right amount of functionality so that the 80%
+cases can be realized and nothing more.  We felt that the essentials boiled down to:
+
+1.  Ability to perform direct expansion of variables in a string
+2.  Ability to specify default values via a prioritized mapping function but without support for
+    defaults as a syntax-level feature
+
+### What should the syntax be?
+
+The exact syntax for variable expansion has a large impact on how users perceive and relate to the
+feature.  We considered implementing a very restrictive subset of the shell `${var}` syntax.  This
+syntax is an attractive option on some level, because many people are familiar with it.  However,
+this syntax also has a large number of lesser known features such as the ability to provide
+default values for unset variables, perform inline substitution, etc.  
+
+In the interest of preventing conflation of the expansion feature in Kubernetes with the shell
+feature, we chose a different syntax similar to the one in Makefiles, `$(var)`. 
```

Comment:
For the sake of posterity, you may want to make a note about why the conflicts with the shell construct $(command) aren't a big deal or how they should be dealt with